MDCStatefulRipple.h

```diff
@@ -0,0 +1,79 @@
+// Copyright 2019-present the Material Components for iOS authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import "MDCRippleView.h"
+
+/**
+ Provides the current state of the ripple. The ripple is either in its normal state, or in the
+ selected state where the ripple remains spread on the view.
+ - MDCRippleStateNormal: The ripple isn't currently presented.
+ - MDCRippleStateHighlighted: The ripple is activated and shown.
+ - MDCRippleStateSelected: The ripple is in the selected state.
+ - MDCRippleStateDragged: The ripple is in the dragged state.
+ */
+typedef NS_OPTIONS(NSInteger, MDCRippleStateNew) {
+  MDCRippleStateNewNormal = 0,
+  MDCRippleStateNewHighlighted = 1 << 0,
+  MDCRippleStateNewSelected = 1 << 1,
+  MDCRippleStateNewDragged = 1 << 2,
+};
+
+@interface MDCStatefulRippleView : MDCRippleView
+
+/**
+ This BOOL is set to YES if the ripple is currently selected, or NO otherwise.
+ It only has significance if selectionMode is activated.
+ Defaults to NO.
+ */
+@property(nonatomic, getter=isSelected) BOOL selected;
+
+/**
+ This BOOL is set to YES if the ripple is currently highlighted, or NO otherwise.
+ Defaults to NO.
+ */
+@property(nonatomic, getter=isRippleHighlighted) BOOL rippleHighlighted;
+
+/**
+ This BOOL is set to YES if the ripple is currently dragged, or NO otherwise.
+ This state is only triggered manually by setting this property to YES.
+ Defaults to NO.
+ */
+@property(nonatomic, getter=isDragged) BOOL dragged;
+
+/**
+ This BOOL is set to YES if the ripple is currently in the selection mode, or NO otherwise.
+ Defaults to NO.
```

Comment:
What is "selection mode"? Who is suppose to set this to YES?

Reply:
selectionMode tells the class that a tap will cause the surface to go into a selected/deselected state. If set to NO (default), a tap will cause a ripple effect that quickly disperses. When set to YES, a tap disperses a ripple that has a selected color and stays without dispersing. this property needs to be set by the component using the rippleView and that is in a selection mode. some ways a user initiates a selection mode is through an affordance in their app or through a long press on a collection of items.

Reply:
What was the decision to differentiate from `allowsSelection` from the conversation above it feels like this API is supposed to do the same thing.

Reply:
I didn't quite understand your comment, but yes it should be on par with UITableView and UICollectionView's allowsSelection. The ripple just needs to be aware of this flag to provide the correct behavior.

Reply:
My suggestion is change:

```diff
- @property(nonatomic) BOOL selectionMode;
+ @property(nonatomic) BOOL allowsSelection;
```

Reply:
> selectionMode tells the class that a tap will cause the surface to go into a selected/deselected state. If set to NO (default), a tap will cause a ripple effect that quickly disperses. When set to YES, a tap disperses a ripple that has a selected color and stays without dispersing. this property needs to be set by the component using the rippleView and that is in a selection mode. some ways a user initiates a selection mode is through an affordance in their app or through a long press on a collection of items.

Tip: Add this explanation to the source file (current comment is a little obscure)